File: 백준/Gold/2470. 두 용액/두 용액.py
N=int(input())
li=[int(x) for x in input().split()]
li.sort()

i=0
j=len(li)-1

cal=[]
m=abs(li[i]+li[j])
m1=li[0]
m2=li[j]

while i<j:
    if li[i]+li[j]>0:
        if m>abs(li[i]+li[j]):
            m=abs(li[i]+li[j])
            m1=li[i]
            m2=li[j]
        j-=1
    elif li[i]+li[j]<0:
        if m>abs(li[i]+li[j]):
            m=abs(li[i]+li[j])
            m1=li[i]
            m2=li[j]
        i+=1
    else:
        m = abs(li[i] + li[j])
        m1 = li[i]
        m2 = li[j]
        break




print(m1,m2)

File: 백준/Gold/2470. 두 용액/두 용액.java
import java.io.*;
import java.util.*;



public class Main {


    public static void main(String[] args) throws NumberFormatException,IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int num=Integer.parseInt(br.readLine());
        int[] arr= new int[num];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for(int i=0;i<num;i++){
            arr[i]=Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);

        int i=0;
        int j=arr.length-1;
        int mim=Integer.MAX_VALUE;
        int m1=0;
        int m2=0;

        while(i<j){
            int sum=arr[i]+arr[j];
            if (Math.abs(sum)<mim){
                mim=Math.abs(sum);
                m1=arr[i];
                m2=arr[j];
            }
            if (sum>0){
                j--;
            }
            else{
                i++;
            }

        }
        System.out.println(m1 + " " + m2);

    }

}